lib/knapsack_pro/extensions/rspec_extension.rb

```diff
@@ -0,0 +1,164 @@
+    #
+    # Often, the structure of the `knapsack__` method is inspired by an existing RSpec method. The `knapsack__` method should have a similar structure to the original RSpec method.
+    # It should be easy to see a code difference when a new RSpec version is released.
+    # Please use permalinks to RSpec source code in your comments when you create a new `knapsack__` method.
+    module RSpecExtension
+      def self.setup!
+        RSpec::Core::World.prepend(World)
+        RSpec::Core::Runner.prepend(Runner)
+        RSpec::Core::Configuration.prepend(Configuration)
+      end
+
+      module World
+        # Based on `announce_filters`
+        # https://github.com/rspec/rspec-core/blob/f8c8880dabd8f0544a6f91d8d4c857c1bd8df903/lib/rspec/core/world.rb#L171
+        def knapsack__setup
+          fail_if_config_and_cli_options_invalid
+        end
+      end
+
+      module Runner
+        def knapsack__mark_setup_as_done
+          @knapsack__setup_done = true
+        end
+
+        def knapsack__setup_done?
+          !!@knapsack__setup_done
+        end
+
+        # Based on:
+        # https://github.com/rspec/rspec-core/blob/f8c8880dabd8f0544a6f91d8d4c857c1bd8df903/lib/rspec/core/runner.rb#L98
+        def knapsack__setup(stream_error = $stderr, stream_out = $stdout)
+          # Abstract from RSpec::Core::Runner#setup, since we do not need to set any filters or files at this point,
+          # and we do not want to let world.announce_filters be called, since it will print
+          # out `No examples found.` message.
+          configure(stream_error, stream_out)
+
+          world.knapsack__setup
+
+          knapsack__mark_setup_as_done
+        end
+
+        def knapsack__wants_to_quit?
+          world.wants_to_quit
+        end
+
+        # `rspec_is_quitting` was added in RSpec 3.11.0
+        # Changelog:
+        #   https://github.com/rspec/rspec-core/blob/f8c8880dabd8f0544a6f91d8d4c857c1bd8df903/Changelog.md#3110--2022-02-09
+        # PR:
+        #   https://github.com/rspec/rspec-core/pull/2926
+        # We support RSpec < 3.11.0.
+        def knapsack__rspec_is_quitting?
+          world.respond_to?(:rspec_is_quitting) && world.rspec_is_quitting
+        end
+
+        def knapsack__exit_early
+          _exit_status = configuration.reporter.exit_early(exit_code)
+        end
+
+        # Based on:
+        # https://github.com/rspec/rspec-core/blob/f8c8880dabd8f0544a6f91d8d4c857c1bd8df903/lib/rspec/core/configuration.rb#L546
+        # Set with --error-exit-code CODE - Override the exit code used when there are errors loading or running specs outside of examples.
+        def knapsack__error_exit_code
+          _default = nil || configuration.error_exit_code
+        end
+
+        def knapsack__deprecated_run_all_when_everything_filtered_enabled?
+          raise "Cannot call the #{__method__} method because setup was not done." unless knapsack__setup_done?
```

Comment:
`@function_core.ensure_no_deprecated_run_all_when_everything_filtered_option!` must be called after `@rspec_runner.knapsack__setup` in the following code:


```ruby
# lib/knapsack_pro/runners/queue/rspec_runner.rb
        def pre_run_setup
           ....
          @rspec_runner.knapsack__setup(@stream_error, @stream_out)

          @function_core.ensure_no_deprecated_run_all_when_everything_filtered_option!(@rspec_runner.knapsack__deprecated_run_all_when_everything_filtered_enabled?)
```

so that the configuration of RSpec is read from the `spec_helper.rb` file first and the `configuration.run_all_when_everything_filtered` value is known .


I removed the following:

```
raise "Cannot call the #{__method__} method because setup was not done." unless knapsack__setup_done?
```

and the integration test fails `bundle exec rspec spec/integration/runners/queue/rspec_runner_spec.rb:1394` when I reverted order of calls:


```ruby
# lib/knapsack_pro/runners/queue/rspec_runner.rb
        def pre_run_setup
           ....
            @function_core.ensure_no_deprecated_run_all_when_everything_filtered_option!(@rspec_runner.knapsack__deprecated_run_all_when_everything_filtered_enabled?)

           @rspec_runner.knapsack__setup(@stream_error, @stream_out)
```

The test failed:

```
  when deprecated run_all_when_everything_filtered option is true
    shows an error message AND sets 1 as exit code (FAILED - 1)
```

so we cover that expectation of using ` @function_core.ensure_no_deprecated_run_all_when_everything_filtered_option!` only after the `  @rspec_runner.knapsack__setup` is called. We could remove the raise.

---

I added the `raise` in the first place to ensure no one is using the `knapsack__deprecated_run_all_when_everything_filtered_enabled` method unless RSpec setup is done. If you would use the `knapsack__deprecated_run_all_when_everything_filtered_enabled` method then it would return false just because the `spec_helper.rb` was not loaded yet (the `@rspec_runner.knapsack__setup` was not called). I added the exception to make it more explicit for developers looking at the code.

What do you think, should we keep or remove the raise?

Reply:
I'd remove it because it adds complexity (with the setup methods and ivar).

It's tested so we won't break it easily. Also, even if we broke it, it wouldn't be a big problem.

If you want to keep a reminder, feel free to add a comment with something like:

```
# Must be called after `knapsack__setup`
def ensure_no_deprecated_run_all_when_everything_filtered_option!
```